drivers: flash: flash_hp_ra: perform blank check before reading

jeremydick opened the pull request:
The value read from unwritten areas of Renesas RAxxx SoCs data flash is undefined. To prevent reading unwritten areas a blank check command is performed first. If the area is blank, we return dummy data so it behaves the same as other flash devices.

de-nordic commented:
@KhiemNguyenT re-assigning to you, your IC.

quytranpzz commented on drivers/flash/Kconfig.renesas_ra:
```diff
@@ -30,4 +30,13 @@ config DUAL_BANK_MODE
 
+config CHECK_BEFORE_READING
+	bool "Verify area before reading it"
+	default y
```
Not all Renesas RA MCU families with flash hp have undefined value after erasing. Can we add `erased-undefined` properties in dts and enable this config based on it? like `default $(dt_nodelabel_bool_prop,flash1,erased-undefined)`

jeremydick replied:
Yes, it looks like the RA4E1, RA4E2, RA4M2, RA4M3, RA6E1, RA6E2, RA6M1, RA6M2, RA6M3, RA6M4, RA6M5, RA8D1, RA8M1, and RA8T1 families have the data flash the will readback an undefined value when erased. Is that all of them, or are there any others I have missed?

quytranpzz replied:
Sorry for late response. There is currently a PR to support the RA4L1 #88036, and the erase value for it is `0xff` ([RA4L1](https://www.renesas.com/en/document/mah/ra4l1-group-users-manual-hardware)), and I think it make more sense for classification

jeremydick replied:
Thanks. Since the erase value for the RA4L1 family is 0xff nothing needs to be done for it, only families that have an undefined erase value need to have the erased-undefined property added

quytranpzz commented on drivers/flash/flash_hp_ra.c:
```diff
@@ -89,9 +144,21 @@ static int flash_ra_read(const struct device *dev, off_t offset, void *data, siz
 	LOG_DBG("flash: read 0x%lx, len: %u", (long)(offset + flash_data->area_address), len);
 
-	memcpy(data, (uint8_t *)(offset + flash_data->area_address), len);
+#ifdef CONFIG_CHECK_BEFORE_READING
```
Can we move the data flash region check here instead of inside the `is_area_readable`. Then, with the code flash region, we won't need to call `is_area_readable`

jeremydick replied:
The flash region check has been moved

quytranpzz commented:
@KhiemNguyenT , @thaoluonguw , @duynguyenxa , @thenguyenyf, @khoa-nguyen-18 
Performing a blank check when the erase value is undefined would improve the current flash driver. Please take a look.

khoa-nguyen-18 reviewed (approved):
LGTM

thenguyenyf commented on drivers/flash/flash_hp_ra.c:
```diff
@@ -31,6 +31,8 @@ BUILD_ASSERT((ERASE_BLOCK_SIZE_1 % FLASH_HP_DF_BLOCK_SIZE) == 0,
 static volatile struct event_flash g_event_flash = {
 	.erase_complete = false,
 	.write_complete = false,
+	.blank = false,
+	.not_blank = false,
```
nit:
```suggestion
#ifdef CONFIG_CHECK_BEFORE_READING
	.blank = false,
	.not_blank = false,
#endif
```

thenguyenyf commented on drivers/flash/flash_hp_ra.c:
```diff
@@ -42,6 +44,10 @@ void bgo_callback(flash_callback_args_t *p_args)
 {
 	if (FLASH_EVENT_ERASE_COMPLETE == p_args->event) {
 		g_event_flash.erase_complete = true;
+	} else if (FLASH_EVENT_BLANK == p_args->event) {
+		g_event_flash.blank = true;
+	} else if (FLASH_EVENT_NOT_BLANK == p_args->event) {
+		g_event_flash.not_blank = true;
 	} else {
```
nit
```suggestion
	}
#ifdef CONFIG_CHECK_BEFORE_READING 
	else if (FLASH_EVENT_BLANK == p_args->event) {
		g_event_flash.blank = true;
	} else if (FLASH_EVENT_NOT_BLANK == p_args->event) {
		g_event_flash.not_blank = true;
	}
#endif 
	else {
```

thenguyenyf commented on drivers/flash/flash_hp_ra.h:
```diff
@@ -97,6 +97,8 @@ struct flash_hp_ra_config {
 struct event_flash {
 	volatile bool erase_complete;
 	volatile bool write_complete;
+	volatile bool blank;
+	volatile bool not_blank;
```
```suggestion
#ifdef CONFIG_CHECK_BEFORE_READING
	volatile bool blank;
	volatile bool not_blank;
#endif
```

thenguyenyf commented on dts/bindings/mtd/renesas,ra-nv-flash.yaml:
```diff
@@ -12,3 +12,7 @@ properties:
     type: boolean
     description: Enable flash programming configuration
+
+  erased-undefined:
```
I'm not sure `erased-undefined` or `blank-check` is better. It seems this prop just to notify the driver that it may need performing a blank check for this flash instance before application read back the flash after erase because of the undefine erase value. The blank check operation should be enabled by set the Kconfig `CHECK_BEFORE_READING=y`.

In this case, I think erased-undefined might make sense.

jeremydick replied:
@soburi I don't think `blank-check` is better since like @thenguyenyf explained, the property is just to inform the driver that the flash does not have a defined value when it is in the erased state. Maybe `erase-value-undefined` would be better for clarity though?

de-nordic commented on drivers/flash/flash_hp_ra.c:
```diff
@@ -89,9 +148,23 @@ static int flash_ra_read(const struct device *dev, off_t offset, void *data, siz
+
+	if (!rc) {
+		memcpy(data, (uint8_t *)(offset + flash_data->area_address), len);
+#ifdef CONFIG_CHECK_BEFORE_READING
+	} else if (rc == -ENODATA) {
+		/* Erased area, return dummy data as an erased page. */
+		memset(data, 0xFF, len);
+		rc = 0;
+#endif
+	}
```
As I have stated, we will not be able to resolve this in this PR, but we need a discussion on -ENODATA as valid return from flash_read.

> Hi @de-nordic, thank you for your feedback.
> 
> In my opinion, returning `-ENODATA` isn’t a typical way for users to check if a flash page has been erased. Most applications, including the `flash/common` test, simply check for `(ret == 0) && (read_buffer_value == 0xFF)` to confirm that the page is erased.

It is not for checking whether flash page is erased, it is rather for checking whether user has placed data in specified location.
Note that there is code in Zephyr that reads page and compares it against erase-value to figure out whether page has been erased (looping through buffer), so it can now be written. In case of your device the code will not be able to recognize a difference between erased area and area that has been written with erase-value. 

> 
> The issue we're facing is that the Renesas flash in some SoC series didn't have the `0xFF` after erase, which causes compatibility problems with this common check. We’ve considered two options to address this:
> 
>     1. Update the Zephyr's flash test/application condition to accommodate different erased values, and make a broad announcement to inform users of Renesas flash drivers that they need to update their condition checks.
> 
>     2. Provide a configuration option in the Renesas flash driver that allows users to choose whether they want the erased buffer to be `0xFF` (as proposed in this PR). If this configuration is not selected, we should return `-ENODATA` (need to update). We believe this makes sense, especially for users porting applications from other MCU vendors to Renesas MCUs who want to avoid making changes to their existing code.

I agree with what you did; what I mean is that we need to clarify definition of flash_read on the matter of returning -ENODATA, modify in tree code (tests, etc) and announce the change, because it is API breaking change. This will take more effort and I do not think it should happen in this PR.

jeremydick replied:
@de-nordic I agree, longer term it makes the most sense to update the API to allow -ENODATA as a valid return for flash_read, and update all usages of the flash driver within the tree handle that.

thenguyenyf reviewed (dismissed):
Please rebase to solve conflict

jeremydick committed:
dts: arm: renesas: Add erase-value-undefined property for Renesas RA flash

Add an erased-undefined property for Renesas RA series MCUS
data flash that will read back undefined values when erased

Signed-off-by: Jeremy Dick <jdick@pivotint.com>

jeremydick committed:
drivers: flash: flash_hp_ra: perform blank check before reading

The value read from unwritten areas of Renesas RAxxx SoCs data
flash is undefined. To prevent reading unwritten areas a
blank check command is performed first. If the area is blank, we
return dummy data so it behaves the same as other flash devices.

Signed-off-by: Jeremy Dick <jdick@pivotint.com>

jeremydick commented:
Rebased to resolve conflicts

khoa-nguyen-18 commented on drivers/flash/Kconfig.renesas_ra:
```diff
@@ -31,4 +31,13 @@ config FLASH_RENESAS_RA_HP_BGO
 	help
 	  Enable Background operations (BGOs)
 
+config CHECK_BEFORE_READING
```
Should we add the `FLASH_RENESAS_RA_HP_` prefix to this config to keep it consistent with the other configs in this file?

khoa-nguyen-18 commented on dts/bindings/mtd/renesas,ra-nv-data-flash.yaml:
```diff
@@ -13,3 +13,7 @@ properties:
     description: |
       Enable data flash programming configuration
+
+  erased-undefined:
```
Could you please rename this property to `erase-value-undefined`, as discussed earlier?

khoa-nguyen-18 reviewed (commented):
I have some minor suggestions regarding your recent update

jeremydick commented:
@KhiemNguyenT could you please review this change to perform a blank check before reading flash with undefined erase values?

KhiemNguyenT reviewed (approved):
LGTM
